L64-python-runtime-proto-parsing.md

```diff
@@ -0,0 +1,136 @@
+gRPC Python Runtime Proto Parsing
+----
+* Author(s): gnossen
+* Approver: lidizheng
+* Status: Draft
+* Implemented in: Python
+* Last updated: April 9, 2020
+* Discussion at: https://groups.google.com/g/grpc-io/c/fuvDt-jZL2k
+
+## Abstract
+
+Generated protobuf code is hard to manage for those without monorepos equipped with a hermetic build system. This has been echoed by the maintainers of libraries wrapping gRPC as well as direct users of gRPC. In this document, we aim to lay out a more Pythonic way of dealing with Protocol Buffer-backed modules.
+
+
+## Background
+
+gRPC and protocol buffers evolved in a Google ecosystem supported by a [monorepo](https://www.wired.com/2015/09/google-2-billion-lines-codeand-one-place/) and [hermetic build system](https://arxiv.org/pdf/1702.01715.pdf). When engineers check a `.proto` file into source control using that toolchain, the resultant code generated by the protocol compiler does not end up in source control. Instead, the code is generated on demand during builds and cached in a distributed build artifact store. While such sophisticated machinery has started to become available in the open source community (e.g. [Bazel](https://bazel.build/)), it has not yet gained much traction. Instead, small Git repos and non-hermetic language-specific build tools are the norm. As a result, code generated by the protocol compiler is often checked into source control alongside the code which uses it.
```

Comment:
>Instead, the code is generated on demand during builds and cached in a distributed build artifact store. While such sophisticated machinery has started to become available in the open source community (e.g. [Bazel](https://bazel.build/)), it has not yet gained much traction. Instead, small Git repos and non-hermetic language-specific build tools are the norm.

I think you need to support this point more. It sounds like you're implying here that the issue is missing technology, but generating protocol files doesn't require Bazel or a monorepo, it just requires a basic build system, no sophisticated machinery; even a basic Makefile can do it; you should more explicit point out why the build systems that open source projects have today cannot support generating protocol files.

Reply:
Other build systems can and (some) do support it ([setuptools](https://github.com/grpc/grpc/blob/e63fa5fdf0fc24bd041ba99d2c58d2974595bf5d/tools/distrib/python/grpcio_tools/grpc_tools/command.py#L51), [Bazel](https://github.com/grpc/grpc/blob/master/bazel/python_rules.bzl#L209)). It's simply a question of effort to implement (in the case that no tooling exists for your particular build system) or effort to integrate (in the case that it does). The weakness that all build systems share is that they live *outside the language*. The strongest use case for this is "running as a script". That is, running a `.py` file invoking a gRPC method without any build system at all (.e.g `python3 grpc_client.py`). In my experience, the vast majority of Python usage falls into that category.  These users *could* integrate with a build system, but, for a simple script, given the choice, even I wouldn't.

I'll clarify in the proposal.